packages/agent/src/oidc-v2.ts

```
@@ -0,0 +1,689 @@
+  const jwtu8a = Convert.string(jwt).toUint8Array();
+  const chacha = xchacha20poly1305(codeChallenge, nonce, additionalData);
+  const ciphertextAndTag = chacha.encrypt(jwtu8a);
+
+  /** The cipher output concatenates the encrypted data and tag
+   * so we need to extract the values for use in the JWE. */
+  const ciphertext = ciphertextAndTag.subarray(0, -16);
+  const authenticationTag = ciphertextAndTag.subarray(-16);
+
+  const compactJwe = [
+    Convert.object(protectedHeader).toBase64Url(),
+    '', // Empty string since there is no wrapped key.
+    Convert.uint8Array(nonce).toBase64Url(),
+    Convert.uint8Array(ciphertext).toBase64Url(),
+    Convert.uint8Array(authenticationTag).toBase64Url(),
+  ].join('.');
+
+  return compactJwe;
+}
+
+// todo: split up oidc and connect specific stuff in the future
+async function createResponseObject(
+  options: RequireOnly<
+    HybridAuthResponse,
+    'iss' | 'sub' | 'aud' | 'delegation_grants' | 'privateKeyJwks'
+  >
+) {
+  const currentTimeInSeconds = Math.floor(Date.now() / 1000);
+
+  const responseObject: HybridAuthResponse = {
+    ...options,
+    iat : currentTimeInSeconds,
+    exp : currentTimeInSeconds + 600, // Expires in 10 minutes.
+  };
+
+  return responseObject;
+}
+
+async function signJwt({
+  did,
+  data,
+}: {
+  did: BearerDid;
+  data: Record<string, unknown>;
+}) {
+  try {
+    const header = Convert.object({
+      alg : 'EdDSA',
+      kid : did.document.verificationMethod![0].id,
+      typ : 'JWT',
+    }).toBase64Url();
+
+    const payload = Convert.object(data).toBase64Url();
+
+    // signs using ed25519 EdDSA
+    const signer = await did.getSigner();
+    const signature = await signer.sign({
+      data: Convert.string(`${header}.${payload}`).toUint8Array(),
+    });
+
+    const signatureBase64Url = Convert.uint8Array(signature).toBase64Url();
+
+    const jwt = `${header}.${payload}.${signatureBase64Url}`;
+
+    return jwt;
+  } catch (e) {
+    console.error(e);
```

Should not eat all errors.

---

I can return an error